chain/network/src/network_protocol/mod.rs

```diff
@@ -52,7 +53,21 @@ pub struct AccountData {
 pub const MAX_ACCOUNT_DATA_SIZE_BYTES: usize = 10000; // 10kB
 
 impl AccountData {
-    pub fn sign(self, signer: &dyn near_crypto::Signer) -> anyhow::Result<SignedAccountData> {
+    /// Serializes AccountData to proto and signs it using `signer`.
+    /// Panics if AccountData.account_id doesn't match signer.validator_id(),
+    /// as this would likely be a bug.
+    /// Returns an error if the serialized data is too large to be broadcasted.
+    /// TODO(gprusak): consider separating serialization from signing (so introducing an
+    /// intermediate SerializedAccountData type) so that sign() then could fail only
```

Comment:
I always view failable serialization with suspicion: if serialization is failable, it raises the question of why we are alowing such potentially invalid values to exist in the first place? 

In this case, I'd probably try to enforce size limit as an invariant of `AccountData` itself (so, eg, make it hash <= 10 peers or something)

But yeah, `serialize()?.sign()?` is also a great general approach!

Reply:
The AccountData is POD struct (i.e. with public fields) and it acts more like a builder. I've put a constraint on the total size of the serialized thing, which is a constraint verifiable once the value is "built" and I don't see anything wrong with that. Here sign() call acts as a constructor of SignedAccountData struct which is more than just a composition of its arguments (i.e. has additional constraints).